File: src/queue/ts/queue.ts
import { CircularQueueProps } from "./types"
import { EmptyQueueError, FullQueueError } from "./utils"

export class CircularQueue<Type> {
    private props: CircularQueueProps<Type>

    constructor(size: number) {
        if (typeof size !== "number") {
            throw new Error("Size must be a number")
        } else if (size <= 0) {
            throw new Error("Size must be greater then 0")
        }

        this.props = {
            end: 0,
            begin: 0,
            size,
            items: [],
            itemCount: 0
        }
    }

    public enqueue(value: Type): void {
        if (this.isFull()) {
            throw new FullQueueError()
        }

        this.props.items[this.props.end] = value
        this.props.end = (this.props.end + 1) % this.props.size
        this.props.itemCount += 1
    }

    public dequeue(): Type {
        if (this.isEmpty()) {
            throw new EmptyQueueError()
        }

        const value = this.props.items[this.props.begin]
        this.props.itemCount -= 1
        this.props.begin = (this.props.begin + 1) % this.props.size

        return value
    }

    public peek(): Type {
        if (this.isEmpty()) {
            throw new EmptyQueueError()
        }
        return this.props.items[this.props.begin]
    }

    public print(): void {
        if (this.isEmpty()) {
            throw new EmptyQueueError()
        }

        let output = ""
        for (let i = 0; i < this.props.itemCount; i++) {
            const index = (this.props.begin + i) % this.props.size
            output += `(${JSON.stringify(this.props.items[index])})-`
        }

        console.log(output.slice(0, -1))
    }

    public isEmpty(): boolean {
        return this.props.itemCount === 0
    }

    public isFull(): boolean {
        return this.props.itemCount === this.props.size
    }
}

File: src/queue/c/queue.c
#include <stdio.h>
#include <stdlib.h>
#include "queue.h"


struct Queue* createQueue(int size){
    if (size <= 0){
        printf("Size must be greater than 0!\n");
        return NULL;
    }

    int *items = malloc(sizeof(int)*size);
    struct Queue *queue = malloc(sizeof(struct Queue));

    queue->items = items;
    queue->itemCount = 0;
    queue->size = size;
    queue->begin = 0;
    queue->end = 0;

    return queue;
}

int enqueue(struct Queue *queue, int value){
    if (isFull(queue)){
        printf("Queue is full!\n");
        return -1;
    }

    queue->items[queue->end] = value;
    queue->itemCount += 1;
    queue->end = (queue->end + 1) % queue->size;

    return 0;
}

int dequeue(struct Queue *queue){
    if (isEmpty(queue)){
        printf("Queue is empty!\n");
        return -1;
    }

    int value = queue->items[queue->begin];
    queue->itemCount -= 1;
    queue->begin = (queue->begin + 1) % queue->size;

    return value;

}

int peek(struct Queue *queue){
    if (isEmpty(queue)){
        printf("Queue is empty!\n");
        return -1;
    }

    return queue->items[queue->begin];
}

int print(struct Queue *queue){
    if (isEmpty(queue)){
        printf("Queue is empty!\n");
        return -1;
    }

    for (int i = 0; i<queue->itemCount; i++){
        int index = (queue->begin + i) % queue->size;
        if (i > 0){
            printf("-");
        }
        printf("(%d)", queue->items[index]);
    }
    printf("\n\n");

    return 0;
}

int isFull(struct Queue *queue){
    return queue->itemCount == queue->size;
}

int isEmpty(struct Queue *queue){
    return queue->itemCount == 0;
}
